# Last-minute calculations for CM Morales

## Option 1: Climate Champion Alternative

- Revise targets using the timelines set out in the January draft of the policy, in which the first BEPS compliance period began in 2027, to include greenhouse gas intensity targets (GHGIT) that are 68% smaller for all intervals and all building types (assuming no exemptions or loopholes, except for affordable and social housing and community nonprofits).
- We estimate this would increase the emissions reductions associated with the policy by approximately 41% from the current June draft, and would result in a 90% reduction in annual emissions by 2040, relative to the 2019 baseline. Am I correct that as the 68% is unchanged from previous modeling, the 41% is good to go as well? **NO, it reduces net emissions from 2027-2050 by an additional 48% compared to the June draft. Overall, it produces 62% less CO2 than the June plan.**
- Additional benefits: 
   - Ramps good union jobs quickly and steadily
   - Avoids potential catastrophic policy failure due to supply chain bottlenecks in final compliance cycles. 
   - Brings life-saving cooling to Seattle homes and workplaces sooner rather than later.

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns

In [7]:
from baseline_model import BaselineBEPSModel

## June proposal vs Jan proposal with 68% reduction

In [3]:
june_no_change = pd.read_csv('june_baseline_8_3.csv')

In [8]:
june_no_change_emissions = june_no_change['compliant_emissions'].sum()

In [19]:
JAN_TIMELINE_PATH = '../data/input_data/scen_1_reformatted.csv'
BUILDING_DATA_PATH = '../data/input_data/building_data_recleaned_8_3.csv'
FINES_YEARS = [2030, 2035, 2040, 2045, 2050]
FINE_PER_SQFT = 2.5

EMISSIONS_PATH = '../data/input_data/energy_emissions.csv'

In [20]:
jan_68_reduction_timeline = pd.read_csv(JAN_TIMELINE_PATH)
jan_68_reduction_timeline['ghgi'] = jan_68_reduction_timeline['ghgi'] * 0.32
jan_68_reduction_timeline.to_csv('jan_68_reduction_timeline.csv')

In [21]:
jan_with_68_red_model = BaselineBEPSModel(EMISSIONS_PATH, 'jan_68_reduction_timeline.csv', BUILDING_DATA_PATH, FINES_YEARS, FINE_PER_SQFT)

In [22]:
jan_with_68_red_model.calculate_baseline_model(2027, 2050)

/Users/ihc/programming/350_seattle/BPS/experiments/benchmarking_models/models/baseline_model.py:69: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  temp_df['year'] = pd.Series([year]*len(temp_df))
/Users/ihc/programming/350_seattle/BPS/experiments/benchmarking_models/models/baseline_model.py:70: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  temp_df['expected_baseline'] = input_df.apply(lambda building: self._get_expected_baseline(building, year), axis=1)
/Users/ihc/programming/350_seattle/BPS/experiments/benc

Model calculations complete. Access the model dataframe as model_name.scenario_results


In [14]:
jan_with_68_red_model.scenario_results.to_csv('jan_with_68_reduction_full_model.csv')

In [26]:
jan_with_68_red_emissions = jan_with_68_red_model.scenario_results['compliant_emissions'].sum()

In [27]:
jan_with_68_red_emissions

1806263161.6562467

In [18]:
june_no_change_emissions

4814919286.478417

In [28]:
# what percent of june emissions is the jan w/ reductions emissions?

(june_no_change_emissions - jan_with_68_red_emissions)/june_no_change_emissions

0.6248611754035592

The January timeline with a 68% reduction in targets produces 62% less CO2 than the June timeline.

## January w/ reductions vs. June when compared with the baseline

In [29]:
JAN_BASELINE_MODEL_PATH = 'jan_baseline_8_3.csv'
jan_baseline = pd.read_csv(JAN_BASELINE_MODEL_PATH)

# emissions in 2026 will be the same as emissions in 2027 w/o improvements to infrastructure
JAN_BASELINE_2026_EMISSIONS_KG = jan_baseline[jan_baseline['year'] == 2027]['expected_baseline'].sum()
JAN_BASELINE_2026_EMISSIONS_KG

459452209.8585

In [30]:
emissions_with_no_improvements_2027_2050 = JAN_BASELINE_2026_EMISSIONS_KG * 24

In [31]:
# reductions with Jan w/ 68% reduction

(emissions_with_no_improvements_2027_2050 - jan_with_68_red_emissions) / emissions_with_no_improvements_2027_2050

0.8361941384672221

In [32]:
# reductions with June plan

(emissions_with_no_improvements_2027_2050 - june_no_change_emissions) / emissions_with_no_improvements_2027_2050

0.5633460180801236

In [34]:
jan_with_68_net_reduction = emissions_with_no_improvements_2027_2050 - jan_with_68_red_emissions
jan_with_68_net_reduction

9220589874.947754

In [35]:
june_net_reduction = emissions_with_no_improvements_2027_2050 - june_no_change_emissions
june_net_reduction

6211933750.125583

In [36]:
jan_with_68_net_reduction / june_net_reduction

1.4843348699205536

January plan with 68% reduction results in a 48% larger reduction in total emissions 2027-2050 than the June plan.

That is calculated by:

`(net_jan_reduction / net_june_reduction) - 1 = 1.48 - 1 = 0.48 = 48%`

I have no idea where they're getting 41%. I suspect it's from doing something like summing the emissions in the 2019 dataset, without adjusting for the changes in emissions factors by 2026.

## Option 2: Climate Leader Alternative

- Revise the policy to use timelines and targets from OSE’s January draft BEPS policy, in which the first BEPS compliance period began in 2027, for publicly owned buildings and large buildings over 90,000 square feet. Maintain current draft targets for all other building types.
- We estimate this would decrease emissions reductions associated with the policy by approximately 20.30% from the current June draft, and would result in a 70.84% reduction in annual emissions by 2040, relative to the 2019 baseline. Would it be easy to get these numbers by using Jan timeline for buildings over 90K sq ft and the June timeline for smaller buildings? **20.2% reduction from June, 71.5% reduction in annual emissions by 2040**
- And I hate to say it, but there's some confusion over the publicly owned buildings part -- would it be reasonably easy to model Jan for >90K sq. ft., June for smaller?

All of these questions will be answered with the following model:
- 90k+ sq ft buildings will use the Jan. plan
- < 90k sq ft buildings will use the June plan
- Public/private status will be ignored

In [54]:
# Combine the Jan model for buildings > 90k sq ft and the June model for <90k sq ft
# A and B buildings are buildings above 90k sq ft in the classification scheme

jan_90k_plus_buildings = jan_baseline[jan_baseline['sq_ft_classification'].isin(['A', 'B'])]

In [44]:
june_less_than_90k_buildings = june_no_change[~june_no_change['sq_ft_classification'].isin(['A', 'B'])]

In [45]:
len(june_less_than_90k_buildings)

65400

In [48]:
option_2_buildings = pd.concat([jan_90k_plus_buildings, june_less_than_90k_buildings], axis=0)

In [50]:
# sanity check

len(option_2_buildings) == len(jan_baseline)

True

What is the emissions for Option 2 vs. the June plan?

In [53]:
(june_no_change_emissions - option_2_buildings['compliant_emissions'].sum()) / june_no_change_emissions

0.20222879058681817

Option 2 leads to a 20.2% reduction in emissions between 2027 and 2050 compared to the June proposal.

What is the emissions in 2040 with this combo vs. the baseline?

In [51]:
option_2_in_2040 = option_2_buildings[option_2_buildings['year'] == 2040]['compliant_emissions'].sum()

In [52]:
option_2_in_2040 / JAN_BASELINE_2026_EMISSIONS_KG

0.29458787648770735

So Option 2 gives us a 71.5% reduction in annual emissions by 2040.

In [55]:
option_2_buildings.to_csv('opt_2_climate_leader_model.csv')

## Visualize it

In [58]:
june_baseline_annual = june_no_change.groupby(by='year').agg({'compliant_emissions': 'sum'})
june_baseline_annual.columns = ['June Proposal']
june_baseline_annual['June Proposal'] = june_baseline_annual['June Proposal'] / 1000.0

In [59]:
option_1_annual = jan_with_68_red_model.scenario_results.groupby(by='year').agg({'compliant_emissions': 'sum'})
option_1_annual.columns = ['Option 1: Climate Champion']
option_1_annual['Option 1: Climate Champion'] = option_1_annual['Option 1: Climate Champion'] / 1000.0

In [60]:
option_2_annual = option_2_buildings.groupby(by='year').agg({'compliant_emissions': 'sum'})
option_2_annual.columns = ['Option 2: Climate Leader']
option_2_annual['Option 2: Climate Leader'] = option_2_annual['Option 2: Climate Leader'] / 1000.0

In [61]:
viz_data = june_baseline_annual.join([option_1_annual, option_2_annual])
viz_data.head()

June Proposal  Option 1: Climate Champion  Option 2: Climate Leader
year                                                                       
2027.0  457844.114281               275171.473448             357699.374971
2028.0  457844.114281               238971.397221             357699.374971
2029.0  457844.114281               211495.943022             357699.374971
2030.0  421976.728557               181963.042537             331154.327551
2031.0  363679.195236               153632.449453             304378.115272

In [ ]:
sns.lineplot(data=viz_data, x='Year', y='Emissions (metric tons)').set(title='Annual emissions (metric tons)')

In [64]:
viz_data.index.astype('int')

Index([2027, 2028, 2029, 2030, 2031, 2032, 2033, 2034, 2035, 2036, 2037, 2038,
       2039, 2040, 2041, 2042, 2043, 2044, 2045, 2046, 2047, 2048, 2049, 2050],
      dtype='int64', name='year')